### PR TITLE
support MiniCPM-Llama3-V 2.5 

### PR DESCRIPTION
**Pull request by irexyc**
## Motivation

support MiniCPM-Llama3-V 2.5 

https://github.com/InternLM/lmdeploy/issues/1693

### Usage

```python
from lmdeploy import pipeline
from lmdeploy.vl import load_image

pipe = pipeline('openbmb/MiniCPM-Llama3-V-2_5')

image = load_image('https://raw.githubusercontent.com/open-mmlab/mmdeploy/main/tests/data/tiger.jpeg')
response = pipe(('describe this image', image))
print(response)
```

### Quantization

`lmdeploy lite auto_awq /nvme/shared/MiniCPM-Llama3-V-2_5/ --work-dir /nvme/shared/MiniCPM-Llama3-V-2_5-4bit/`

### CODE REVIEW

File: docs/en/multi_modal/minicpmv.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# minicpmv
```
Comment (lvhan028) [inline docs/en/multi_modal/minicpmv.md:1]: Let's use `MiniCPM-V` as the title

File: docs/zh_cn/multi_modal/minicpmv.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# minicpmv
```
Comment (lvhan028) [inline docs/zh_cn/multi_modal/minicpmv.md:1]: Let's use MiniCPM-V as the title

File: docs/zh_cn/multi_modal/minicpmv.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+## 简介
+
+[MiniCPM-V](https://github.com/OpenBMB/MiniCPM-V) 是面向图文理解的端侧多模态大模型系列。该系列模型接受图像和文本输入，并提供高质量的文本输出。 LMDeploy 已在 TurboMind 后端支持 MiniCPM-Llama3-V-2_5 模型 [openbmb/MiniCPM-Llama3-V-2_5](https://huggingface.co/openbmb/MiniCPM-Llama3-V-2_5)
```
Comment (lvhan028) [inline docs/zh_cn/multi_modal/minicpmv.md:5]: LMDeploy 支持了 [MiniCPM-Llama3-V-2_5](https://huggingface.co/openbmb/MiniCPM-Llama3-V-2_5) 模型，通过 TurboMind 引擎推理

File: lmdeploy/turbomind/deploy/converter.py [hunk 4]
```diff
@@ -23,6 +23,7 @@
     'internlm2': 'internlm2',
     'xcomposer2': 'xcomposer2',
     'deepseekvl': 'deepseekvl',
+    'minicpmv': 'minicpmv',
```
Comment (lvhan028) [inline lmdeploy/turbomind/deploy/converter.py:26]: May update after PR #1702

File: lmdeploy/vl/templates.py [hunk 5]
```diff
@@ -239,6 +239,35 @@ def append_image_token(self, prompt, num_images: int):
         return res
 
 
+class MiniCPMVTempateWrapper(VLChatTemplateWrapper):
```
Comment (lvhan028) [inline lmdeploy/vl/templates.py:242]: What's the chat template of the LLM model in MiniCPM?

**Comment by QwertyJack** [comment]
How to do the turbomind format conversion?

**Comment by benihime91** [comment]
hey tried out this PR, i am getting the following error :
`TypeError: MiniCPMV.forward() missing 1 required positional argument: 'data'` (stack pasted below)
Error is happening with both `PytorchEngineConfig` & `TurbomindEngineConfig`.

Also this is how pipeline is initialized in my code
```python
if args.backend == "torch":
    backend_config = PytorchEngineConfig(tp=args.tp, session_len=args.session_len)
elif args.backend == "turbomind":
    backend_config = TurbomindEngineConfig(tp=args.tp, session_len=args.session_len)
else:
    backend_config = None
gen_config = GenerationConfig(
    temperature=args.temperature,
    top_p=args.top_p,
    max_new_tokens=args.max_new_tokens,
    random_seed=args.seed,
)
pipe = pipeline(args.model, backend_config=backend_config)
```

==== STACKTRACE ====
```
Traceback (most recent call last):
  File "/mnt/data1/ayushman/projects/dash-diffusion/third_party/lmdeploy/lmdeploy/pytorch/engine/request.py", line 17, in _raise_exception_on_finish
    task.result()
  File "/mnt/data1/ayushman/projects/dash-diffusion/third_party/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 824, in async_loop
    await __step(True)
  File "/mnt/data1/ayushman/projects/dash-diffusion/third_party/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 810, in __step
    raise e
  File "/mnt/data1/ayushman/projects/dash-diffusion/third_party/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 802, in __step
    raise out
  File "/mnt/data1/ayushman/projects/dash-diffusion/third_party/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 751, in _async_loop_background
    await self._async_step_background(
  File "/mnt/data1/ayushman/projects/dash-diffusion/third_party/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 659, in _async_step_background
    output = await self._async_model_forward(inputs,
  File "/mnt/data1/ayushman/projects/dash-diffusion/third_party/lmdeploy/lmdeploy/utils.py", line 250, in __tmp
    return (await func(*args, **kwargs))
  File "/mnt/data1/ayushman/projects/dash-diffusion/third_party/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 576, in _async_model_forward
    return await __forward(inputs)
  File "/mnt/data1/ayushman/projects/dash-diffusion/third_party/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 554, in __forward
    return await self.model_agent.async_forward(
  File "/mnt/data1/ayushman/projects/dash-diffusion/third_party/lmdeploy/lmdeploy/pytorch/engine/model_agent.py", line 739, in async_forward
    output = self._forward_impl(inputs,
  File "/mnt/data1/ayushman/projects/dash-diffusion/third_party/lmdeploy/lmdeploy/pytorch/engine/model_agent.py", line 705, in _forward_impl
    output = model_forward(
  File "/mnt/data1/ayushman/projects/dash-diffusion/third_party/lmdeploy/lmdeploy/pytorch/engine/model_agent.py", line 481, in model_forward
    output = patched_model.patched_forward(
  File "/mnt/data1/ayushman/projects/dash-diffusion/third_party/lmdeploy/lmdeploy/pytorch/models/patch.py", line 242, in __call__
    output = self._model(*args, **kwargs)
  File "/home/ayushman/miniforge3/envs/lmdeploy/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/ayushman/miniforge3/envs/lmdeploy/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
TypeError: MiniCPMV.forward() missing 1 required positional argument: 'data'
```

File: lmdeploy/vl/templates.py (hunk 5 above)
Comment (irexyc) [reply lmdeploy/vl/templates.py:242]: llama3

**Comment by irexyc** [comment]
> How to do the turbomind format conversion?

We don't support convert the vlm to turbomind format. Instead, we support load the model in huggingface format.

**Comment by irexyc** [comment]
@benihime91

Currently, this model only support turbomind backend. According to your log, you are using the pytorch backend, please post the log of turbomind backend.

**Comment by HaoLiuHust** [comment]
Thanks for your work,  I have tried the PR, the quant step successed, but when load the mode, an error occured:
```
Traceback (most recent call last):
  File "/mnt/pai-storage-8/jieshen/code/lmdeploy/demo.py", line 4, in <module>
    pipe = pipeline('./MiniCPM-Llama3-V-2_5-4bit')
  File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/api.py", line 89, in pipeline
    return pipeline_class(model_path,
  File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/serve/vl_async_engine.py", line 21, in __init__
    super().__init__(model_path, **kwargs)
  File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/serve/async_engine.py", line 206, in __init__
    self._build_pytorch(model_path=model_path,
  File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/serve/async_engine.py", line 268, in _build_pytorch
    self.engine = Engine(model_path=model_path,
  File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 135, in __init__
    self.model_agent = AutoModelAgent.from_pretrained(
  File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/pytorch/engine/model_agent.py", line 610, in from_pretrained
    return build_model_agent(pretrained_model_name_or_path,
  File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/pytorch/engine/model_agent.py", line 1271, in build_model_agent
    model_agent = BaseModelAgent(model_path,
  File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/pytorch/engine/model_agent.py", line 638, in __init__
    self.patched_model = self._build_model(
  File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/pytorch/engine/model_agent.py", line 657, in _build_model
    hf_model = AutoModelForCausalLM.from_pretrained(
  File "/data/miniconda3/lib/python3.9/site-packages/transformers/models/auto/auto_factory.py", line 558, in from_pretrained
    return model_class.from_pretrained(
  File "/data/miniconda3/lib/python3.9/site-packages/transformers/modeling_utils.py", line 3640, in from_pretrained
    hf_quantizer.preprocess_model(
  File "/data/miniconda3/lib/python3.9/site-packages/transformers/quantizers/base.py", line 182, in preprocess_model
    return self._process_model_before_weight_loading(model, **kwargs)
  File "/data/miniconda3/lib/python3.9/site-packages/transformers/quantizers/quantizer_awq.py", line 85, in _process_model_before_weight_loading
    model, has_been_replaced = replace_with_awq_linear(
  File "/data/miniconda3/lib/python3.9/site-packages/transformers/integrations/awq.py", line 178, in replace_with_awq_linear
    _, has_been_replaced = replace_with_awq_linear(
  File "/data/miniconda3/lib/python3.9/site-packages/transformers/integrations/awq.py", line 178, in replace_with_awq_linear
    _, has_been_replaced = replace_with_awq_linear(
  File "/data/miniconda3/lib/python3.9/site-packages/transformers/integrations/awq.py", line 178, in replace_with_awq_linear
    _, has_been_replaced = replace_with_awq_linear(
  [Previous line repeated 2 more times]
  File "/data/miniconda3/lib/python3.9/site-packages/transformers/integrations/awq.py", line 165, in replace_with_awq_linear
    model._modules[name] = target_cls(
  File "/data/miniconda3/lib/python3.9/site-packages/awq/modules/linear/gemm.py", line 101, in __init__
    assert self.in_features % self.group_size == 0
AssertionError
```
the demo code is:
```
from lmdeploy import pipeline
from lmdeploy.vl import load_image
from lmdeploy import pipeline, TurbomindEngineConfig

engine_config = TurbomindEngineConfig(model_format='awq')

pipe = pipeline('./MiniCPM-Llama3-V-2_5-4bit',backend_config=engine_config)

image = load_image('https://raw.githubusercontent.com/open-mmlab/mmdeploy/main/tests/data/tiger.jpeg')
response = pipe(('describe this image', image))
print(response)
```

**Review by lvhan028 — APPROVED** [review]

**Comment by HaoLiuHust** [comment]
```


> Thanks for your work, I have tried the PR, the quant step successed, but when load the mode, an error occured:
> 
> ```
> Traceback (most recent call last):
>   File "/mnt/pai-storage-8/jieshen/code/lmdeploy/demo.py", line 4, in <module>
>     pipe = pipeline('./MiniCPM-Llama3-V-2_5-4bit')
>   File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/api.py", line 89, in pipeline
>     return pipeline_class(model_path,
>   File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/serve/vl_async_engine.py", line 21, in __init__
>     super().__init__(model_path, **kwargs)
>   File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/serve/async_engine.py", line 206, in __init__
>     self._build_pytorch(model_path=model_path,
>   File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/serve/async_engine.py", line 268, in _build_pytorch
>     self.engine = Engine(model_path=model_path,
>   File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 135, in __init__
>     self.model_agent = AutoModelAgent.from_pretrained(
>   File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/pytorch/engine/model_agent.py", line 610, in from_pretrained
>     return build_model_agent(pretrained_model_name_or_path,
>   File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/pytorch/engine/model_agent.py", line 1271, in build_model_agent
>     model_agent = BaseModelAgent(model_path,
>   File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/pytorch/engine/model_agent.py", line 638, in __init__
>     self.patched_model = self._build_model(
>   File "/mnt/pai-storage-8/jieshen/code/lmdeploy/lmdeploy/pytorch/engine/model_agent.py", line 657, in _build_model
>     hf_model = AutoModelForCausalLM.from_pretrained(
>   File "/data/miniconda3/lib/python3.9/site-packages/transformers/models/auto/auto_factory.py", line 558, in from_pretrained
>     return model_class.from_pretrained(
>   File "/data/miniconda3/lib/python3.9/site-packages/transformers/modeling_utils.py", line 3640, in from_pretrained
>     hf_quantizer.preprocess_model(
>   File "/data/miniconda3/lib/python3.9/site-packages/transformers/quantizers/base.py", line 182, in preprocess_model
>     return self._process_model_before_weight_loading(model, **kwargs)
>   File "/data/miniconda3/lib/python3.9/site-packages/transformers/quantizers/quantizer_awq.py", line 85, in _process_model_before_weight_loading
>     model, has_been_replaced = replace_with_awq_linear(
>   File "/data/miniconda3/lib/python3.9/site-packages/transformers/integrations/awq.py", line 178, in replace_with_awq_linear
>     _, has_been_replaced = replace_with_awq_linear(
>   File "/data/miniconda3/lib/python3.9/site-packages/transformers/integrations/awq.py", line 178, in replace_with_awq_linear
>     _, has_been_replaced = replace_with_awq_linear(
>   File "/data/miniconda3/lib/python3.9/site-packages/transformers/integrations/awq.py", line 178, in replace_with_awq_linear
>     _, has_been_replaced = replace_with_awq_linear(
>   [Previous line repeated 2 more times]
>   File "/data/miniconda3/lib/python3.9/site-packages/transformers/integrations/awq.py", line 165, in replace_with_awq_linear
>     model._modules[name] = target_cls(
>   File "/data/miniconda3/lib/python3.9/site-packages/awq/modules/linear/gemm.py", line 101, in __init__
>     assert self.in_features % self.group_size == 0
> AssertionError
> ```
> 
> the demo code is:
> 
> ```
> from lmdeploy import pipeline
> from lmdeploy.vl import load_image
> from lmdeploy import pipeline, TurbomindEngineConfig
> 
> engine_config = TurbomindEngineConfig(model_format='awq')
> 
> pipe = pipeline('./MiniCPM-Llama3-V-2_5-4bit',backend_config=engine_config)
> 
> image = load_image('https://raw.githubusercontent.com/open-mmlab/mmdeploy/main/tests/data/tiger.jpeg')
> response = pipe(('describe this image', image))
> print(response)
> ```

已解决，原因是使用了本地的lmdeploy源码，而不是编译后安装的包，在其他目录运行就正常了

**Comment by insogao** [comment]
server api支持吗？

**Comment by QwertyJack** [comment]
> server api支持吗？

支持，参考[支持模型](https://github.com/InternLM/lmdeploy/blob/main/docs/zh_cn/supported_models/supported_models.md#turbomind-%E6%94%AF%E6%8C%81%E7%9A%84%E6%A8%A1%E5%9E%8B)。

**Pull request merged**